source/vtk_utils.cc

```diff
@@ -0,0 +1,546 @@
+    dof_handler.distribute_dofs(fe_system);
+    DoFRenumbering::block_wise(dof_handler);
+
+    // Read all data into output_vector
+    output_vector.reinit(dof_handler.n_dofs());
+    unsigned int dof_offset = 0;
+    unsigned int field_idx  = 0;
+    // Cell data
+    for (int i = 0; i < cell_data->GetNumberOfArrays(); ++i, ++field_idx)
+      {
+        vtkDataArray *arr = cell_data->GetArray(i);
+        if (!arr)
+          continue;
+        vtkIdType n_tuples = arr->GetNumberOfTuples();
+        int       n_comp   = arr->GetNumberOfComponents();
+        for (vtkIdType j = 0; j < n_tuples; ++j)
+          for (int k = 0; k < n_comp; ++k)
+            output_vector[dof_offset + j * n_comp + k] =
+              arr->GetComponent(j, k);
+        dof_offset += n_tuples * n_comp;
+      }
+    // Point data
+    for (int i = 0; i < point_data->GetNumberOfArrays(); ++i, ++field_idx)
+      {
+        vtkDataArray *arr = point_data->GetArray(i);
+        if (!arr)
+          continue;
+        vtkIdType n_tuples = arr->GetNumberOfTuples();
+        int       n_comp   = arr->GetNumberOfComponents();
+        for (vtkIdType j = 0; j < n_tuples; ++j)
+          for (int k = 0; k < n_comp; ++k)
+            output_vector[dof_offset + j * n_comp + k] =
+              arr->GetComponent(j, k);
+        dof_offset += n_tuples * n_comp;
+      }
+  }
+
+  template <int dim>
+  void
+  fill_distributed_vector_from_serial(
+    const IndexSet       &owned_dofs,
+    const Vector<double> &serial_vec,
+    const std::map<Point<dim>, types::global_dof_index, PointComparator<dim>>
+                                               &serial_map,
+    LinearAlgebra::distributed::Vector<double> &parallel_vec,
+    const std::map<Point<dim>, types::global_dof_index, PointComparator<dim>>
+            &parallel_map,
+    MPI_Comm comm)
+  {
+    Assert(parallel_vec.size() == 0,
+           ExcMessage("The parallel vector must be empty before filling it."));
+    AssertThrow(owned_dofs.n_elements() > 0,
+                ExcMessage("The owned DoF index set must not be empty."));
+    // Initialize parallel layout of the vector using DoFHandler
+    parallel_vec.reinit(owned_dofs, comm);
+
+    // Transfer data from serial to parallel vector
+    for (const auto &p_pair : parallel_map)
+      {
+        const auto &pt             = p_pair.first;
+        const auto &parallel_index = p_pair.second;
+
+        if (!owned_dofs.is_element(parallel_index))
+          continue;
+
+        auto it = serial_map.find(pt);
+        if (it != serial_map.end())
+          {
+            types::global_dof_index serial_index = it->second;
+            parallel_vec[parallel_index]         = serial_vec[serial_index];
+          }
+        else
+          {
+            std::cerr << "No match found for point: " << pt << std::endl;
+            AssertThrow(false, ExcInternalError());
+          }
+      }
+
+    parallel_vec.compress(VectorOperation::insert);
+  }
```

Comment:
This will only work for a scalar serial_vec. In case of vector valued components, serial_vec will contain more dofs that point to the same support point, so the logic used here cannot be used in the general case. 

We should input a DoFHandler, the serial_vec, and a single map from distributed to serial vertex index. Then we should use `dist_id = cell->vertex_index(i)`, and if it is locally owned, set `dist_vector[cell->vertex_dof_index(i, comp) = serial_vec[n_comps*dist_id + comp]` for each comp in n_comps.